connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala

```diff
@@ -312,14 +368,11 @@ private[sql] class AvroSerializer(
    */
   private def resolveAvroType(avroType: Schema): (Boolean, Schema) = {
     if (avroType.getType == Type.UNION) {
-      val fields = avroType.getTypes.asScala
-      val actualType = fields.filter(_.getType != Type.NULL)
-      if (fields.length != 2 || actualType.length != 1) {
-        throw new UnsupportedAvroTypeException(
-          s"Unsupported Avro UNION type $avroType: Only UNION of a null type and a non-null " +
-            "type is supported")
+      nonNullUnionBranches(avroType) match {
+        case Seq() => (true, Schema.create(Type.NULL))
+        case Seq(singleType) => (true, singleType)
+        case _ => (false, avroType)
```

Comment:
`avroTypes` contains the null type in the union. Should this return the union types with the null filtered out? That is what it previously did for nullable unions.

Reply:
I think this is exactly what we are doing here, by using the output of `nonNullUnionBranches`.
We make however an exception of `["null"]`.

Reply:
So we only make an exception for _optional types_.  For _complex unions_, we however need the null further down.